doc/man/radzap.8
.\" $Id$
.ds RD /usr/local/etc/raddb
.ds RL /usr/adm
.ds RP /usr/local/etc/raddb
.TH radzap 8 "March 10, 2000" "FSF"
.SH NAME
radzap - delete Radius login record.

.SH SYNOPSIS
\fBradzap\fR
[\fB-c\fR|\fB-q\fR]
[\fB-h\fR]
[\fB-n\fR NAS]
[\fB-p\fR port]
username

.SH DESCRIPTION
\fBradzap\fR searches the Radius log files for matching login records
and closes them.
.PP
At least one of \fB-n\fR, \fB-p\fR options or username must be
specified. If they are used in conjunction, they are taken as if
joined by logical AND operation.

.SH OPTIONS
.IP -c
Enable confirm mode. In this mode \fBradzap\fR asks for a confirmation
before zapping each matching record. Every line beginning with a `y'
is taken as positive response, otherwise it is taken as negative
response.
.IP -q
Disable confirm mode.
.IP -h
Display short help summary and exit.
.IP -n
Specify NAS name to zap user from.
.IP -p
Specify the port number of the session to be zapped. The port number
can be specified either in its full form, i.e ``radzap -p S02'' or in
its short form, i.e. ``radzap -p 2''.

.SH ENVIRONMENT VARIABLE
.IP RADZAP_CONFIRM
Specifies the default behaviour of the \fBradzap\fR. When set to a
positive value, \fBradzap\fR starts in confirm mode.

.SH FILES
.IP \*(RL/radwtmp
.IP \*(RL/radutmp

.SH BUGS
The information in this manpage may be obsolete or incomplete. Please
refer to texinfo documentation for full information about GNU Radius
tools.

.SH AUTHORS
Sergey Poznyakoff, gray@farlep.net

.SH SEE ALSO
radiusd(8rad), radwho(1rad)


NEWS
GNU Radius NEWS -- history of user-visible changes. 2004-04-23
Copyright (C) 2002,2003,2004 Free Software Foundation, Inc.
See the end of file for copying conditions.

Please send radius bug reports to <bug-gnu-radius@gnu.org>
^L
Version 1.2.3:

* libgnuradius

This is a library of functions for creation, handling and sending
requests using RADIUS protocol.

All programs have been rewritten to link with libgnuradius. On most
sites this will mean linking against a shared library, which will
reduce the size of the executables.

* gnuradius.scm

This is a guile module allowing to use libgnuradius functions. It
supercedes radscm program, which has been removed.

* Rewrite language

** Added i18n support

** New built-in functions:

*** Functions to access internal fields of a RADIUS request.
*** Interfaces to the Radius NAS database (raddb/naslist).
*** Interfaces to DNS lookup functions.

* radiusd

** Improved checking for multiple logins. Previous versions relied entirely
on the contents of /var/log/radutmp file. Starting at this version,
radiusd offers at least two methods of checking for multiple logins:
using the traditional radutmp file and using the SQL database. New
keywords has been added to the sqlserver file that declare the SQL
queries to be used when retrieving the information about currently
active sessions.

More methods of checking will be added in future versions.

** When an unsupported authentication type is requested, radiusd first
checks if an extension Scheme module is provided that handles that
authentication type. If such module is found, it is invoked to handle
the authentication.

** System accounting can be turned off by specifying `system no;' in
`acct' section of raddb/config. 

** New configuration statement 'load-module' allows to load arbitrary
Scheme modules.

** Detailed log filenames are configurable via `detail-file-name'
statements in `auth' and `acct' sections of raddb/config.

** Support for Guile versions prior to 1.6 has been withdrawn.

* Bugfixes

** Allow to omit port numbers in `listen' statements (raddb/config), as
described in the documentation.
    


Version 1.2:

* Important notes for upgrading from earlier versions of GNU Radius

** Upgrade your dictionaries.

This version does a very strict semantic checking on the dictionary
files. Old dictionaries may produce lots of warnings and
errors, therefore you are advised to upgrade them. To do so,
run `make update-dictionary'. It will update your main 
(raddb/dictionary) and secondary dictionaries (raddb/dict/*),
preserving any changes you did to the former. 

Notice, that, starting from this version, the default main dictionary
file will not contain any attribute declarations, instead it will
merely include the necessary dictionaries. However, you may add your
declarations to it and modify it to your linking.

Any modifications to the dictionary files in raddb/dict are now
strongly discouraged, as these files will be updated by each new
release. Put your modifications to the main dictionary file only.
This file will remain unchanged during subsequent installs.

Two new statements are provided to help you customize your dictionary.
See section `** Dictionary' below.

** Update your raddb/config and raddb/rewrite

The rewrite procedures are no longer stored in a single file
`radiusd/rewrite', instead they may occupy several source file.
The rewrite procedures shipped with this version of GNU Radius
are installed in ${prefix}/share/radius/1.2/rewrite. We strongly
advise you to add the following statement to your raddb/config:

rewrite {
        load "checknas.rw";
        load "log-hook.rw";
};

You may then remove the old raddb/rewrite file, *unless* you have
modified it.

If you have modified it, follow the procedure described below:

1. Load this file, by adding the following statement to the above
`rewrite' block:

	load "rewrite";

2. Make sure to remove from this file or rename the following functions:
check_snmp_u, check_snmp_s, decimal, check_max_finger, check_unix.

** Tailor log file format.

The format of log file is now fully customizable via a special Rewrite
procedures, called "hooks" (search for logging-hook in the text below
for the detailed description). To revert to logfile format used in
versions 1.0-1.1, add the following two statements to your raddb/config,
right after `logging {' line:

       	prefix-hook "compat_log_prefix";
	suffix-hook "compat_log_suffix";

** Using Guile extensions

If you are using Guile extensions, notice that the default location
for Radius scheme modules is now ${prefix}/share/radius/1.2/scheme.
	
* New features:

** Support for Status-Server request.

** Support for Tunnel Protocol (RFC 2865)

** The main configuration file raddb/config allows to specify several
Rewrite source files to be loaded instead of the single predefined
file raddb/rewrite.

** Arbitrary Rewrite expressions may be used in A/V pairs, for example:

DEFAULT	     Auth-Type = SQL
	     Framed-IP-Address =
	           "=10.10.10.1 + %[NAS-Port-Id]/16 + %[NAS-Port-Id]"

** Test shell mode considerably improved.

** The internal attributes do not appear in detailed logs. The
special flag 'l' has been added to the dictionary syntax to
turn this feature off. Thus, for example:

ATTRIBUTE Orig-User-Name  2101	string	- [L--R-R]l

will make Orig-User-Name attribute visible in detailed log files.

** Changes to Rewrite language

*** usual dotted-quad notation may be used to represent IP numbers.

*** a number of new built-in functions is added

*** #pragma regexp allows to alter the type of regular expressions
used. The rest of line following `#pragma regexp' must consist of
whitespace-separated regexp modifiers. Each regexp modifier is one
of the following keywords:

  extended      Use extended regular expressions
  icase         Ignore case
  newline       Match-any-character operators don't match a newline.

optionally preceded by plus, indicating enabling the feature, or

radzap/radzap.c
/* This file is part of GNU Radius.
   Copyright (C) 2000, 2001, 2002, 2003, 2004, 2005, 2007, 2008, 2010,
   2013 Free Software Foundation, Inc.

   Written by Sergey Poznyakoff

   GNU Radius is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.
  
   GNU Radius is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with GNU Radius.  If not, see <http://www.gnu.org/licenses/>. */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <sys/types.h>
#include <sys/time.h>
#include <sys/file.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <errno.h>
#include <netinet/in.h>

#include <common.h>
#include <radius/radargp.h>
#include <radius/radutmp.h>

#define LOCK_LEN sizeof(struct radutmp)

int confirm_flag;

struct arguments {
	char *radutmp;
        char *user;
        char *nas;
        int port;
};
const char *argp_program_version = "radzap (" PACKAGE ") " VERSION;
static char doc[] = N_("delete Radius login records");

static struct argp_option options[] = {
        {NULL, 0, NULL, 0,
         N_("radzap specific switches:"), 0},
        {"confirm", 'c', NULL, 0,
         N_("ask for confirmation before zapping"), 0},
        {"log-directory", 'l', "DIR", 0,
         N_("set logging directory"), 0},
        {"file", 'f', N_("FILE"), 0,
         N_("operate on FILE instead of /var/log/radutmp"), 0},
        {"nas", 'n', N_("NASNAME"), 0,
         N_("zap user from given NAS"), 0},
        {"port", 'p', N_("NUMBER"), 0,
         N_("zap user coming from given port"), 0},
        {"quiet", 'q', NULL, 0,
         N_("do not ask for confirmation before zapping"), 0},
        {NULL, 0, NULL, 0, NULL, 0}
};

static error_t
parse_opt(int key, char *arg, struct argp_state *state)
{
        struct arguments *args = state->input;
        
        switch (key) {
        case 'c':
                confirm_flag = 1;
                break;
        case 'l':
                grad_log_dir = arg;
                break;
        case 'f':
                args->radutmp = arg;
                break;
        case 'n':
                args->nas = arg;
                break;
        case 'p':
                if (*arg == 's' || *arg == 'S')
                        ++arg;
                args->port = atoi(arg);
                break;
        case 'q':
                confirm_flag = 0;
                break;
        case ARGP_KEY_ARG:
                args->user = state->argv[state->next - 1];
                break;
        case ARGP_KEY_FINI:
                if (!args->user && !args->nas && args->port == -1) {
                        grad_log(GRAD_LOG_ERR,
                                 _("at least one port, nas or user must be specified"));
                        exit(1);
                }
                break;
        default:
                return ARGP_ERR_UNKNOWN;
        }
        return 0;
}

static struct argp argp = {
        options,
        parse_opt,
        NULL,
        doc,
        grad_common_argp_child,
        NULL, NULL
};


/*
 *      Zap a user, or all users on a NAS, from the radutmp file.
 */
int
radzap(grad_netdef_t *netdef, int port, char *user, time_t t)
{
        struct radutmp  *up;
        radut_file_t    file;
        grad_uint32_t   netaddr;

        if (t == 0) 
                time(&t);

        if ((file = grad_ut_setent(grad_utmp_file, 0)) == NULL) {
                grad_log(GRAD_LOG_ERR|GRAD_LOG_PERROR, 
                         _("can't open file %s"), grad_utmp_file);
                exit(1);
        }       
        /*
         *      Find the entry for this NAS / portno combination.
         */
        while (up = grad_ut_getent(file)) {
                if ((netdef
		     && !grad_ip_in_net_p(netdef, htonl(up->nas_address)))
		    || (port >= 0 && port != up->nas_port)
		    || (user != NULL && strcmp(up->login, user))!= 0
		    ||  up->type != P_LOGIN) {
                        continue;
                }
                if (!confirm(up))
                        continue;

                up->type = P_IDLE;
                up->time = t;
                grad_ut_putent(file, up);
                write_wtmp(up);
        }
        grad_ut_endent(file);

        return 0;
}

int
confirm(struct radutmp *utp)
{
        char buf[GRAD_MAX_LONGNAME];
        grad_nas_t *cl;
        char *s = NULL;
        
        if (cl = grad_nas_lookup_ip(ntohl(utp->nas_address)))
                s = cl->shortname;
        if (s == NULL || s[0] == 0) 
                s = grad_ip_gethostname(ntohl(utp->nas_address),
					buf, sizeof(buf));
        
        printf(_("radzap: zapping %s from %s, port %d"),
               utp->login,
               s,
               utp->nas_port);
        if (confirm_flag) {
                printf(": Ok?");
                fgets(buf, sizeof(buf), stdin);
                if (buf[0] != 'y' && buf[0] != 'Y') {
                        printf(_("Not confirmed\n"));
                        return 0;
                } else
                        return 1;
        } 
        printf("\n");
        return 1;
}

int
write_wtmp(struct radutmp *ut)
{
        return grad_radwtmp_putent(grad_wtmp_file, ut);
}

/*
 *      Zap a user from the radutmp and radwtmp file.
 */
int
main(int argc, char **argv)
{
        grad_netdef_t netdef, *netdefp = NULL;
        time_t  t;
        char    *path;
        char *s;        
        struct arguments args;
        
        grad_app_setup();

        if (s = getenv("RADZAP_CONFIRM"))
                confirm_flag = atoi(s);
	args.radutmp = NULL;
        args.user = NULL;
        args.nas  = NULL;
        args.port = -1;
        if (grad_argp_parse(&argp, &argc, &argv, 0, NULL, &args))
                return 1;
	if (args.radutmp)
		grad_utmp_file = args.radutmp;

        /*
         *      Read the "naslist" file.
         */
        path = grad_mkfilename(grad_config_dir, RADIUS_NASLIST);
        if (grad_nas_read_file(path) < 0)
                exit(1);
        grad_free(path);

        if (args.nas) {
                grad_nas_t *np;
                np = grad_nas_lookup_name(args.nas);
		if (np) {
                        netdefp = &np->netdef;
		} else {
			if (grad_ip_getnetaddr(args.nas, &netdef)) {
				fprintf(stderr,
					_("%s: host not found.\n"), args.nas);
				return 1;
			}
			netdefp = &netdef;
                } 
        }
                
        t = time(NULL);
        radzap(netdefp, args.port, args.user, t);
        return 0;
}

